pkg/tools/nmap.go

```diff
@@ -28,22 +38,84 @@ var NmapToolDefinition = llms.FunctionDefinition{
 type NmapTool struct{}
 
 type NmapToolArgs struct {
-	IP string `json:"ip"`
+	IP   string   `json:"ip"`
+	Args []string `json:"args,omitempty"`
+}
+
+type PortInfo struct {
+	Port    string
+	State   string
+	Service string
 }
 
 // Call executes the command with the given arguments.
 func (s NmapTool) Call(ctx context.Context, input string) (string, error) {
 	nmapToolArgs := NmapToolArgs{}
 	if err := json.Unmarshal([]byte(input), &nmapToolArgs); err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to unmarshal JSON: %w", err)
+	}
+
+	if nmapToolArgs.IP == "" || net.ParseIP(nmapToolArgs.IP) == nil {
+		return "", fmt.Errorf("invalid or missing IP: %q", nmapToolArgs.IP)
+	}
+
+	var args []string
+	if len(nmapToolArgs.Args) > 0 {
+		args = append([]string{}, nmapToolArgs.Args...)
+		ipPresent := false
+		for _, a := range args {
+			if a == nmapToolArgs.IP {
+				ipPresent = true
+				break
+			}
+		}
+		if !ipPresent {
+			args = append(args, nmapToolArgs.IP)
+		}
+	} else {
+		args = []string{
+			"-v",
+			"-T4",
+			"-PA",
+			"-sV",
+			"--version-all",
+			"--osscan-guess",
+			"-A",
+			"-sS",
+			"-p", "1-65535",
+			nmapToolArgs.IP,
+		}
 	}
-	cmd := exec.Command("nmap", "-v", "-T4", "-PA", "-sV", "--version-all", "-osscan-guess", "-A", "-sS", "-p", "1-65535", nmapToolArgs.IP)
 
+	cmd := exec.Command("nmap", args...)
 	output, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to execute nmap: %w - output: %s", err, string(output))
+	}
+
+	ports := parseNmapPorts(string(output))
+	msfQueries := GenerateMsfQueries(ports)
```

Comment:
still not related to nmap

nmap tool just outputs the opened ports

metasploit tool, or even shared metasploit related util package should generate needed queries from the ports somewhere outside nmap

at least mark that in TODO